react/utils/slots.tsx

```diff
@@ -0,0 +1,167 @@
+  treePath: string
+  runtime: RenderContext
+  hydration: Hydration
+  baseTreePath: string
+}
+
+interface ListContentQuery {
+  listContent: Array<{ contentJSON: string }>
+}
+
+export function generateSlot({
+  treePath,
+  slotName,
+  slotValue,
+  hydration,
+}: GenerateSlotArgs) {
+  const hasSlotMarker = treePath.indexOf('slot') > -1
+  const treePathWithoutSlotMarker = hasSlotMarker
+    ? treePath.substring(0, treePath.indexOf('-slot'))
+    : treePath
+  const newTreePath = `${treePathWithoutSlotMarker}/${slotValue}`
+
+  const SlotComponent: FC<any> = memo(props => {
+    const runtime = useRuntime()
+    let extension = runtime.extensions[newTreePath]
+
+    const slotChildren = getChildExtensions(runtime, newTreePath)
+    let componentProps = extension?.props ?? {}
+
+    if (props.id) {
+      const hasLabel = slotValue.includes('#')
+      const dynamicTreePath = `${newTreePath}${hasLabel ? '-' : '#'}slot${
+        props.id
+      }`
+
+      if (!runtime.extensions[dynamicTreePath]) {
+        componentProps = extension?.props ?? {}
+        const baseSlotBlockId = extension?.blockId
+        const dynamicSlotProps = { ...props, ...componentProps }
+
+        return (
+          <DynamicSlot
+            treePath={dynamicTreePath}
+            runtime={runtime}
+            hydration={hydration}
+            component={extension?.component ?? null}
+            props={dynamicSlotProps}
+            blockId={baseSlotBlockId}
+            baseTreePath={newTreePath}
+          >
+            {slotChildren}
+          </DynamicSlot>
+        )
+      }
+
+      extension = runtime.extensions[dynamicTreePath]
+      componentProps = extension?.props ?? {}
+    }
+
+    const extensionContent = extension?.content
+    const componentLoaderPropsWithContent = {
+      ...props,
+      ...componentProps,
+      ...extensionContent,
+    }
+
+    return (
+      <ComponentLoader
+        component={extension?.component ?? null}
+        props={componentLoaderPropsWithContent}
+        treePath={newTreePath}
+        runtime={runtime}
+        hydration={hydration}
+      >
+        {slotChildren}
+      </ComponentLoader>
+    )
+  })
+
+  SlotComponent.displayName = `${slotName}Slot`
+
+  return SlotComponent
+}
+
+const DynamicSlot: FC<DynamicSlotProps> = ({
+  treePath,
+  runtime,
+  hydration,
+  component,
+  props,
+  children,
+  blockId,
+  baseTreePath,
+}) => {
+  const { data, loading, error } = useQuery<ListContentQuery>(ListContent, {
+    variables: {
+      bindingId: runtime.binding?.id,
+      template: runtime.page,
+      blockId,
+      treePath: treePath.replace(runtime.page, '*'),
+      pageContext: runtime.route.pageContext,
+    },
+  })
+
+  let extensionContent =
+    (data?.listContent[1]?.contentJSON &&
+      (JSON.parse(data.listContent[1]?.contentJSON) as Record<string, any>)) ??
+    (data?.listContent[0]?.contentJSON &&
+      (JSON.parse(data.listContent[0]?.contentJSON) as Record<string, any>))
```

Comment:
wat

Reply:
Woould it be possible to assign some of these values to variable? This is kind of cryptic to people from outside

Reply:
Sure! I agree 😂